File: functions/private/gof.m
%
% GOF Goodness of Fit statistics
%
%   stats = GOF(V,Vfit,Ndof,sigma)
%   Given some input data (V) and its corresponding fit (Vfit), with (Ndof)
%   degrees of freedom and with an underlying error with standard deviation
%   (sigma), returns a structure (stats) with multiple statistical
%   indicators of goodness of fit.
%

% This file is a part of DeerLab. License is MIT (see LICENSE.md).
% Copyright(c) 2019-2020: Luis Fabregas, Stefan Stoll and other contributors.

function stats = gof(V,Vfit,Ndof,sigma)

if nargin<4
    sigma = std(V - Vfit);
end
    
% Ensure correct dimensionality
V = V(:);
Vfit = Vfit(:);

% Get number of variables
N = numel(V);
% Extrapolate number of parameters
Q = Ndof - N;

% Reduced Chi-squared test
stats.chi2red = 1/Ndof*norm(V - Vfit)^2/sigma^2;

% R-squared test
stats.R2 = 1 - sum((V-Vfit).^2)/sum((Vfit-mean(Vfit)).^2);

% Root-mean square deviation
stats.RMSD = sqrt(sum((V-Vfit).^2)/N);

loglike = N*log(norm(V - Vfit)^2/N);
% Akaike information criterion
stats.AIC =  loglike + 2*Q;

% Corrected Akaike information criterion
stats.AICc = loglike + 2*Q + 2*Q*(Q+1)/(N-Q-1);

% Bayesian information criterion
stats.BIC =  loglike + Q*log(N);

end

File: deerlab/utils/gof.py
# gof.py - Goodness-of-fit statistics
# -----------------------------------
# This file is a part of DeerLab. License is MIT (see LICENSE.md).
# Copyright(c) 2019-2021: Luis Fabregas, Stefan Stoll and other contributors.

import numpy as np
import deerlab as dl

def goodness_of_fit(x,xfit,Ndof,sigma='auto'):
    r""" 
    Goodness of Fit statistics
    ==========================

    Computes multiple statistical indicators of goodness of fit.

    Usage:
    ------
        stats = goodness_of_fit(x,xfit,Ndof)

    Arguments: 
    ----------
    x (N-element array)
        Original data
    xfit (N-element array)
        Fit
    Ndog (scalar, int)
        Number of degrees of freedom
    sigma (scalar, optional)
        Standard dexiation of the noise in x.

    Returns:
    --------
    stats (dict)
        Statistical indicators:
            stats['chi2red'] - Reduced chi-squared
            stats['rmsd'] - Root mean-squared dexiation
            stats['R2'] - R-squared test
            stats['aic'] - Akaike information criterion
            stats['aicc'] - Corrected Akaike information criterion
            stats['bic'] - Bayesian information criterion

    """
    if sigma=='auto':
        sigma = dl.noiselevel(x,'der')

    # Get number of xariables
    N = len(x)
    # Extrapolate number of parameters
    Q = Ndof - N

    # Reduced Chi-squared test
    chi2red = 1/Ndof*np.linalg.norm(x - xfit)**2/sigma**2

    # R-squared test
    R2 = 1 - np.sum((x-xfit)**2)/np.sum((xfit-np.mean(xfit))**2)

    # Root-mean square dexiation
    rmsd = np.sqrt(np.sum((x-xfit)**2)/N)

    # Log-likelihood
    loglike = N*np.log(np.linalg.norm(x - xfit)**2/N)

    # Akaike information criterion
    aic =  loglike + 2*Q

    # Corrected Akaike information criterion
    aicc = loglike + 2*Q + 2*Q*(Q+1)/(N-Q-1)

    # Bayesian information criterion
    bic =  loglike + Q*np.log(N)

    return {'chi2red':chi2red,'R2':R2,'rmsd':rmsd,'aic':aic,'aicc':aicc,'bic':bic}